src/rtpa-client.1
.\" ##########################################################################
.\" ####                                                                  ####
.\" ####                      RTP Audio Server Project                    ####
.\" ####                    ============================                  ####
.\" ####                                                                  ####
.\" #### RTP Audio Client Manpage                                         ####
.\" ####                                                                  ####
.\" ####           Copyright (C) 1999-2026 by Thomas Dreibholz            ####
.\" ####                                                                  ####
.\" #### Contact:                                                         ####
.\" ####    EMail: thomas.dreibholz@gmail.com                             ####
.\" ####    WWW:   https://www.nntb.no/~dreibh/rtpaudio                   ####
.\" ####                                                                  ####
.\" #### ---------------------------------------------------------------- ####
.\" ####                                                                  ####
.\" #### This program is free software: you can redistribute it and/or    ####
.\" #### modify it under the terms of the GNU General Public License as   ####
.\" #### published by the Free Software Foundation, either version 3 of   ####
.\" #### the License, or (at your option) any later version.              ####
.\" ####                                                                  ####
.\" #### This program is distributed in the hope that it will be useful,  ####
.\" #### but WITHOUT ANY WARRANTY; without even the implied warranty of   ####
.\" #### MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    ####
.\" #### GNU General Public License for more details.                     ####
.\" ####                                                                  ####
.\" #### You should have received a copy of the GNU General Public        ####
.\" #### License along with this program.  If not, see                    ####
.\" #### <http://www.gnu.org/licenses/>.                                  ####
.\" ####                                                                  ####
.\" ##########################################################################
.\"
.\" ###### Setup ############################################################
.Dd November 29, 2024
.Dt rtpa-client 1
.Os rtpa-client
.\" ###### Name #############################################################
.Sh NAME
.Nm rtpa-client
.Nd RTP Audio Client (text version)
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm rtpa-client
.Op URL
.Op Fl debug
.Op Fl null
.Op Fl encoding=number
.Op Fl prefix=name
.Op Fl info=infostring
.Op Fl force-ipv4
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm rtpa-client
is the RTP Audio client application.
.Pp
.\" ###### Arguments ########################################################
.Sh OPTIONS
The following options may be provided:
.Bl -tag -width indent
.It URL
.It Fl debug
TBD.
.El
.\" ###### Arguments ########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It rtpa-client rtpa://server.domain/Test1.list
.It rtpa-client rtpa+udp://server.domain/Test1.list
.It rtpa-client rtpa+sctp://server.domain/Test1.list
.El
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz
.br
https://www.nntb.no/~dreibh/rtpaudio
.br
mailto://thomas.dreibholz@gmail.com
.br


ChangeLog
commit 6f741a1851b4ce35a2dc7dbf7f20aac2b716111c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 14:17:00 2023 +0100

    Updated packaging configuration.

commit 01ff4b24a92dc6518e8bdda5621420c0e18b33eb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 14:07:46 2023 +0100

    FreeBSD build fix.

commit ce4414517294195ef6578d8b66ef76babebe92ef
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 13:47:27 2023 +0100

    Packaging update.

commit 65eaa6e46d8b547a6536aeb0adb50446842eb8c5
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 12:32:29 2023 +0100

    Improved FreeBSD test script.

commit 935cee09e85c735aba38035ad3935824be1c0f4d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 10:29:54 2023 +0100

    Updated LSM entry.

commit f4f6bb87107bd8bf6c179c28e1dae1e50eb37e2e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Dec 7 10:20:27 2023 +0100

    Updated packaging configuration.

commit dac93f6bb5bd8923b4156ea87ea89890f9363630
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 17:18:26 2023 +0100

    Updated LSM entry.

commit 0fca1af2d6163cefedf9c5e74cd3241a06492720
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 16:54:41 2023 +0100

    New release rtpaudio-2.0.7.

commit 09a51260dfbf72af1f0fea2962e867e047f17e69
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 13:59:49 2023 +0100

    Added lintian override.

commit 037dbc4ead53e77329f1740b744173c387cf9855
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Dec 6 13:49:44 2023 +0100

    FreeBSD build fix.

commit f8f3d56922297b3b3bd50fc40760593930bb012a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 11:42:31 2023 +0100

    Dependency improvements.

commit 4bf74c50c0891e39d54cd5705852bf76bc4b34fa
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 10:53:51 2023 +0100

    Packaging clean-ups.

commit a55536dd6f962e04d15ecfd0e19f610311d533b7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 09:52:23 2023 +0100

    Updated build scripts

commit 24b6acf828c23e20549cd63a45401a08fbf55936
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Nov 15 09:33:15 2023 +0100

    Updated build scripts

commit fa13ea861b4ea0f939cd26f55f95bd6855b53ec1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 17:50:51 2023 +0100

    Workflows update.

commit 801227ed26423d950eb4ee3fbb9f6a83dbb08a84
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:27:16 2023 +0100

    Minor manpage updates.

commit 22e407d96e0d93c1ff1988004b59c62f03a02abd
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:22:36 2023 +0100

    Header updates.

commit 59469efa6aa8aaa3399f97042936b41fa07d3775
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:18:09 2023 +0100

    Updated CI scripts

commit ec0a4ecaa4aefa6f2f0e5827affa44626b0766a2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Nov 11 16:18:08 2023 +0100

    Updated build scripts

commit 5bec64693ac502fe6f8c4ac42e9d7cacfbf148c8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 13:23:07 2023 +0100

    URL change from UDE to NNTB.

commit c4cfbd10269fb83ba954693d61a1bd62d97db79b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 12:13:30 2023 +0100

    Updated autogen.sh.

commit 81cebd8e6549ed09947ac103379a16ed9385f6da
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 11:55:31 2023 +0100

    Updated build scripts

commit bcd562c874e2730fec7363db09545076a630558f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 11:52:40 2023 +0100

    Updated CI scripts

commit 4776b568eeb87dd91be1216940918536da0c053d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 2 11:52:34 2023 +0100

    Travis CI is dead. Removed obsolete Travis CI scripts and configuration.

commit b2328c6151bb9605cdeae292136ab1d7bda13940
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sun Sep 11 13:47:02 2022 +0200

    New release rtpaudio-2.0.6.

commit ce043eb413eae3a35979e495e421b1ed2f778b43
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:47:32 2022 +0200

    Removed Vcs-* fields from debian/control.

commit c02b46d03953abd5b0e1017e1bb5ef9ae0f54762
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:21:06 2022 +0200

    RPM fix.

commit 3e20567e717c452ff28e0731dbdaf084516322c6
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:19:33 2022 +0200

    Updated CI scripts

commit 641a5f665964ef8bdb0b8c662c2de185f565822a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Aug 31 11:19:32 2022 +0200

    Updated build scripts

commit 4d4c7a4591572142414b8df7af5194abb8b1c06f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu May 5 11:41:26 2022 +0200

    Fix for RPM spec: rtpaudio-all is build now.

commit 43f37568b3ea9828919bcd070a07c7a6db14fec6
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 17 12:38:36 2022 +0100

    Updated LSM entry.

commit 671bc93f3285dbb823c3910798a36dc0deac0552
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 17 12:25:22 2022 +0100

    New release rtpaudio-2.0.5.

commit 44618678ebf444e8f63968ae0ef2a5370ce6b38b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 17 11:30:59 2022 +0100

    Version bump.

commit 827afc69b6a195dca87cd5b7b258e13490136452
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Feb 16 15:52:58 2022 +0100

    Fedora 36+ build fix for libmpegsound.

commit 7a4d6708b485e765412d4f6c0e80a54df17c8769

src/rtpa-client.cc
// ##########################################################################
// ####                                                                  ####
// ####                      RTP Audio Server Project                    ####
// ####                    ============================                  ####
// ####                                                                  ####
// #### Audio Client                                                     ####
// ####                                                                  ####
// ####           Copyright (C) 1999-2026 by Thomas Dreibholz            ####
// ####                                                                  ####
// #### Contact:                                                         ####
// ####    EMail: thomas.dreibholz@gmail.com                             ####
// ####    WWW:   https://www.nntb.no/~dreibh/rtpaudio                   ####
// ####                                                                  ####
// #### ---------------------------------------------------------------- ####
// ####                                                                  ####
// #### This program is free software: you can redistribute it and/or    ####
// #### modify it under the terms of the GNU General Public License as   ####
// #### published by the Free Software Foundation, either version 3 of   ####
// #### the License, or (at your option) any later version.              ####
// ####                                                                  ####
// #### This program is distributed in the hope that it will be useful,  ####
// #### but WITHOUT ANY WARRANTY; without even the implied warranty of   ####
// #### MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    ####
// #### GNU General Public License for more details.                     ####
// ####                                                                  ####
// #### You should have received a copy of the GNU General Public        ####
// #### License along with this program.  If not, see                    ####
// #### <http://www.gnu.org/licenses/>.                                  ####
// ####                                                                  ####
// ##########################################################################


#include "multiaudiowriter.h"
#include "audiodevice.h"
#include "audiodebug.h"
#include "audionull.h"
#include "tdsocket.h"
#include "audioclient.h"
#include "breakdetector.h"
#include "audioquality.h"
#include "strings.h"

#include <assert.h>
#include <fstream>


// Fast Break: Disable break detector to debug thread deadlocks
// #define FAST_BREAK


// Globals
AudioWriterInterface* audioOutput = NULL;
AudioClient*          client      = NULL;
std::ofstream*        gpScript    = NULL;
std::ofstream*        gpData      = NULL;


// ###### Clean up ##########################################################
void cleanUp(const cardinal exitCode = 0)
{
   if(client != NULL)
      delete client;
   if(audioOutput != NULL)
      delete audioOutput;
   if(exitCode == 0)
      std::cout << "Terminated!" << std::endl;
   exit(exitCode);
}


// ###### Initialize GNUplot output streams #################################
void initGNUplot(const char*    prefix,
                 const char*    info,
                 const cardinal layers,
                 const String&  address)
{
   const String scriptName = String(prefix) + ".gp";
   const String dataName   = String(prefix) + ".dat";
   time_t timeStamp = getMicroTime() / 1000000;
   const struct tm* timeptr = localtime(&timeStamp);
   char str[256];
   strftime((char*)&str,256,"%A, %B %d %G %T %Z",timeptr);
   String infoString = "Generated on " + (String)str;
   if(info) {
      infoString = infoString + " " + (String)info;
   }

   gpScript = new std::ofstream(scriptName.getData());
   if(gpScript != NULL) {
      if(gpScript->good()) {
         gpData = new std::ofstream(dataName.getData());
         if(gpData != NULL) {
            if(gpData->good()) {
               *gpScript << "set terminal postscript enhanced color dashed" << std::endl;
               *gpScript << "set encoding iso_8859_1" << std::endl;
               for(cardinal i = 1;i < 33;i++) {
                  *gpScript << "set linestyle " << i << " lt " << i << " lw 5" << std::endl;
               }
               *gpScript << "set grid" << std::endl;

               *gpScript << "set title \"RTP Audio Client " << address << " - Bytes Received\" \"Times-Roman,14\"" << std::endl;
               *gpScript << "set timestamp \"" << infoString << "\" 0,0 \"Times-Roman,12\"" << std::endl;
               *gpScript << "set ylabel \"Bytes [1]\"" << std::endl;
               *gpScript << "set xlabel \"Time [s]\"" << std::endl;
               *gpScript << "plot \"" << dataName << "\" using 1:"
                         << 1 + (layers + 1) << " title "
                         << "\"Total\" with lines ls 1";
               for(cardinal i = 0;i < layers;i++) {
                  *gpScript << ", \"" << dataName << "\" using 1:"
                            << 1 + (i + 1) << " title "
                            << "\"Layer #" << i << "\" with lines ls " << (i + 2);
               }
               *gpScript << std::endl;

               *gpScript << "set title \"RTP Audio Client " << address << " - Packets Received\" \"Times-Roman,14\"" << std::endl;
               *gpScript << "set timestamp \"" << infoString << "\" 0,0 \"Times-Roman,12\"" << std::endl;
               *gpScript << "set ylabel \"Packets [1]\"" << std::endl;
               *gpScript << "set xlabel \"Time [s]\"" << std::endl;
               *gpScript << "plot \"" << dataName << "\" using 1:"
                         << 1 + 2 * (layers + 1) << " title "
                         << "\"Total\" with lines ls 1";
               for(cardinal i = 0;i < layers;i++) {
                  *gpScript << ", \"" << dataName << "\" using 1:"
                            << 1 + (layers + 1) + (i + 1) << " title "
                            << "\"Layer #" << i << "\" with lines ls " << (i + 2);
               }
               *gpScript << std::endl;

               *gpScript << "set title \"RTP Audio Client " << address << " - Packets Lost\" \"Times-Roman,14\"" << std::endl;
               *gpScript << "set timestamp \"" << infoString << "\" 0,0 \"Times-Roman,12\"" << std::endl;
               *gpScript << "set ylabel \"Packets Lost [1]\"" << std::endl;
               *gpScript << "set xlabel \"Time [s]\"" << std::endl;
               *gpScript << "plot \"" << dataName << "\" using 1:"
                         << 1 + 3 * (layers + 1) << " title "
                         << "\"Total\" with lines ls 1";
               for(cardinal i = 0;i < layers;i++) {
                  *gpScript << ", \"" << dataName << "\" using 1:"
                            << 1 + 2 * (layers + 1) + (i + 1) << " title "
                            << "\"Layer #" << i << "\" with lines ls " << (i + 2);
               }
               *gpScript << std::endl;

               *gpScript << "set title \"RTP Audio Client " << address << " - Fraction Lost\" \"Times-Roman,14\"" << std::endl;
               *gpScript << "set timestamp \"" << infoString << "\" 0,0 \"Times-Roman,12\"" << std::endl;
               *gpScript << "set ylabel \"Fraction Lost [%]\"" << std::endl;
               *gpScript << "set xlabel \"Time [s]\"" << std::endl;
               *gpScript << "plot \"" << dataName << "\" using 1:"
                         << 1 + 4 * (layers + 1) << " title "
                         << "\"Total\" with lines ls 1";
               for(cardinal i = 0;i < layers;i++) {
                  *gpScript << ", \"" << dataName << "\" using 1:"
                            << 1 + 3 * (layers + 1) + (i + 1) << " title "
                            << "\"Layer #" << i << "\" with lines ls " << (i + 2);
               }
               *gpScript << std::endl;

               *gpScript << "set title \"RTP Audio Client " << address << " - Interarrival Jitter\" \"Times-Roman,14\"" << std::endl;
               *gpScript << "set timestamp \"" << infoString << "\" 0,0 \"Times-Roman,12\"" << std::endl;
               *gpScript << "set ylabel \"Interarrival Jitter [us]\"" << std::endl;
               *gpScript << "set xlabel \"Time [s]\"" << std::endl;
               *gpScript << "plot \"" << dataName << "\" using 1:"
                         << 1 + 5 * (layers + 1) << " title "
                         << "\"Average\" with lines ls 1";
               for(cardinal i = 0;i < layers;i++) {
                  *gpScript << ", \"" << dataName << "\" using 1:"
                            << 1 + 4 * (layers + 1) + (i + 1) << " title "
                            << "\"Layer #" << i << "\" with lines ls " << (i + 2);
               }
               *gpScript << std::endl;
               return;
            }
            else {
               std::cerr << "ERROR: Unable to create data file " << dataName << "!" << std::endl;
            }
            delete gpData;
         }
      }
      else {
         std::cerr << "ERROR: Unable to create script file " << scriptName << "!" << std::endl;
      }
      delete gpScript;
   }
   exit(1);
}


// ###### Main program ######################################################
int main(int argc, char* argv[])
{
   // ===== Check arguments =================================================
   if(argc < 2) {
      std::cerr << "Usage: " << argv[0] << std::endl
           << "[URL] {[+/-]debug} {[+/-]null} {[+/-]device} {-encoding=number} {-prefix=name} {-info=infostring} {-force-ipv4}" << std::endl;
      exit(0);
   }
   bool            optAudioDevice = true;
   bool            optAudioDebug  = false;
   bool            optAudioNull   = false;
   bool            optForceIPv4   = false;
   const char*     info           = "";
   cardinal        encoding       = 0;
   char*           prefix         = NULL;
   integer         rate           = AudioQuality::HighestSamplingRate;
   integer         bits           = AudioQuality::HighestBits;
   bool            stereo         = true;
   for(cardinal i = 1;i < (cardinal)argc;i++) {
      if(!(strcasecmp(argv[i],"+debug")))               optAudioDebug  = 1;
      else if(!(strcasecmp(argv[i],"+null")))           optAudioNull   = 1;
      else if(!(strcasecmp(argv[i],"+device")))         optAudioDevice = 1;
      else if(!(strcasecmp(argv[i],"-debug")))          optAudioDebug  = 0;
      else if(!(strcasecmp(argv[i],"-null")))           optAudioNull   = 0;
      else if(!(strcasecmp(argv[i],"-device")))         optAudioDevice = 0;
      else if(!(strcasecmp(argv[i],"-force-ipv4")))     optForceIPv4  = true;
      else if(!(strcasecmp(argv[i],"-use-ipv4")))       optForceIPv4  = false;
      else if(!(strncasecmp(argv[i],"-prefix=",8)))     prefix        = &argv[i][8];
      else if(!(strncasecmp(argv[i],"-info=",6)))       info          = &argv[i][6];
      else if(!(strncasecmp(argv[i],"-rate=",6)))       rate          = atol(&argv[i][6]);
      else if(!(strncasecmp(argv[i],"-bits=",6)))       bits          = atol(&argv[i][6]);
      else if(!(strcasecmp(argv[i],"-stereo")))         stereo        = true;
      else if(!(strcasecmp(argv[i],"-mono")))           stereo        = false;
      else if(!(strncasecmp(argv[i],"-encoding=",10)))  encoding      = atol(&argv[i][10]);
      else if(argv[i][0] == '-') {
         std::cerr << "Wrong parameter: " << argv[i] << std::endl;
      }
   }
   if(optForceIPv4) {
      if(InternetAddress::UseIPv6 == true) {
         InternetAddress::UseIPv6 = false;
         std::cerr << "NOTE: IPv6 support disabled!" << std::endl;
      }
   }


   // ====== Initialize audio output device =================================
   MultiAudioWriter* audioOutput = new MultiAudioWriter();
   assert(audioOutput != NULL);
   if(optAudioDebug) {
      AudioDebug* audioDebug = new AudioDebug();
      assert(audioDebug != NULL);
      audioOutput->addWriter(audioDebug);
   }
   if(optAudioNull) {
      AudioNull* audioNull = new AudioNull();
      assert(audioNull != NULL);
      audioOutput->addWriter(audioNull);
   }
   if(optAudioDevice) {
      AudioDevice* audioDevice = new AudioDevice();
      assert(audioDevice != NULL);
      if(!audioOutput->ready()) {
         std::cerr << "WARNING: Unable to open audio device!" << std::endl;
      }
      audioOutput->addWriter(audioDevice);
   }


   // ====== Initialize Audio client ========================================
   client = new AudioClient(audioOutput);
   if(client == NULL) {
      std::cerr << "ERROR: Client::main() - Out of memory!" << std::endl;
      cleanUp(1);
   }


   // ====== Set encoding and quality =======================================
   client->setEncoding(encoding);
   for(cardinal i = 0;i < AudioQuality::ValidRates;i++) {
      if(rate >= AudioQuality::ValidRatesTable[i]) {
         client->setSamplingRate(AudioQuality::ValidRatesTable[i]);
      }
   }
   for(cardinal i = 0;i < AudioQuality::ValidBits;i++) {
      if(bits >= AudioQuality::ValidBitsTable[i]) {
         client->setBits(AudioQuality::ValidBitsTable[i]);
      }
   }
   client->setChannels((stereo == true) ? 2 : 1);


   // ====== Start playing ==================================================
   const bool ok = client->play(argv[1]);
   if(ok == false) {
      std::cerr << "ERROR: Client::main() - AudioClient::play() failed!" << std::endl;
      cleanUp(1);
   }
#ifndef FAST_BREAK
   installBreakDetector();
#endif

   const cardinal layers = 3; // ??????
   if(prefix != NULL) {
      initGNUplot(prefix,info,layers,client->getOurAddressString(InternetAddress::PF_Full));
   }


   // ====== Print status ===================================================
   std::cout << "RTP Audio Client - Copyright (C) 1999-2026 Thomas Dreibholz" << std::endl
             << "-----------------------------------------------------------" << std::endl
             << std::endl
             << "Server Address:  " << client->getServerAddressString() << std::endl;
   char str[32];
   snprintf((char*)&str,sizeof(str),"$%08x",client->getOurSSRC());
   std::cout << "Client SSRC:     " << str     << std::endl
             << "Media URL:       " << argv[1] << std::endl
             << "Layers:          " << layers  << std::endl
             << std::endl;


   // ====== Main loop ======================================================
   cardinal EOFRepeatDelay = 0;
   card64   oldBytes[layers];
   card64   oldPackets[layers];
   card64   oldPacketsLost[layers];
   for(cardinal i = 0;i < layers;i++) {
      oldBytes[i] = oldPackets[i] = oldPacketsLost[i] = 0;
   }
   card64   startTS    = getMicroTime() - 1;
   card64   oldTS      = startTS;         // avoid now=oldTS (-> divide by 0)
   while(client->playing()) {
      // ====== Check for errors ============================================
      const card8 error = client->getErrorCode();
      switch(error) {
         case ME_NoError:
         case ME_NoMedia:
          break;
         case ME_EOF:
            EOFRepeatDelay++;
            if(EOFRepeatDelay > 15) {
               EOFRepeatDelay = 0;
               client->setPosition(0);
            }
          break;
         default:
            std::cerr << std::endl << "Encoder error while playing: #" << (int)error << "." << std::endl;
            goto cleanUp;
          break;
      }
#ifndef FAST_BREAK
      if(breakDetected()) {
         goto cleanUp;
      }
#endif


      // ====== Print information ===========================================
      const card64 position = client->getPosition();
      const card64 seconds  = position / PositionStepsPerSecond;
      card64 bytes[layers];
      card64 packets[layers];
      card64 packetsLost[layers];
      double fractionLost[layers];
      double jitter[layers];
      for(cardinal i = 0;i < layers;i++) {
         bytes[i]        = client->getBytesReceived(i);
         packets[i]      = client->getPacketsReceived(i);
         packetsLost[i]  = client->getPacketsLost(i);
         fractionLost[i] = client->getFractionLost(i);
         jitter[i]       = client->getJitter(i);
      }
      const cardinal currentLayers = client->getLayers();
      const card64 now = getMicroTime();
      if(!optAudioDebug) {
         printf("\x0d%2u:%02u.%02u   [Quality: %d Hz / %d Bit / %s]  [%s]      ",
                (unsigned int)(seconds / 60), (unsigned int)(seconds % 60),
                (unsigned int)((position % PositionStepsPerSecond) / (PositionStepsPerSecond / 100)),
                client->getSamplingRate(),
                client->getBits(),
                ((client->getChannels() == 2) ? "Stereo" : "Mono"),
                client->getEncoding());
         fflush(stdout);
      }


      // ====== Write GNUplot output ========================================
      if((gpData != NULL) && (currentLayers > 0)) {
         card64 oldBytesTotal = 0;
         card64 bytesTotal    = 0;
         *gpData << (double)(now - startTS) / 1000000.0 << " ";
         for(cardinal i = 0;i < layers;i++) {
            oldBytesTotal += oldBytes[i];
            bytesTotal    += bytes[i];
            const card64 bps = (bytes[i] - oldBytes[i]) * 1000000 / (now - oldTS);
            *gpData << bps << " ";
         }
         *gpData << (bytesTotal - oldBytesTotal) * 1000000 / (now - oldTS) << " ";

         card64 oldPacketsTotal = 0;
         card64 packetsTotal    = 0;
         for(cardinal i = 0;i < layers;i++) {
            oldPacketsTotal += oldPackets[i];
            packetsTotal    += packets[i];
            const card64 pps = (packets[i] - oldPackets[i]) * 1000000 / (now - oldTS);
            *gpData << pps << " ";
         }
         *gpData << (packetsTotal - oldPacketsTotal) * 1000000 / (now - oldTS) << " ";

         card64 oldPacketsLostTotal = 0;
         card64 totalPacketsLost    = 0;
         for(cardinal i = 0;i < layers;i++) {
            oldPacketsLostTotal += oldPacketsLost[i];
            totalPacketsLost += packetsLost[i];
            *gpData << (packetsLost[i] - oldPacketsLost[i]) << " ";
         }
         *gpData << (totalPacketsLost - oldPacketsLostTotal) << " ";

         double avgFractionLost = 0.0;
         for(cardinal i = 0;i < layers;i++) {
            avgFractionLost += fractionLost[i];
            *gpData << 100.0 * fractionLost[i] << " ";
         }
         avgFractionLost /= (double)currentLayers;
         *gpData << avgFractionLost * 100.0 << " ";

         double avgJitter = 0.0;
         for(cardinal i = 0;i < layers;i++) {
            avgJitter += jitter[i];
            *gpData << jitter[i] << " ";
         }
         avgJitter /= (double)currentLayers;
         *gpData << avgJitter;

         *gpData << std::endl;
      }


      // ====== Wait ========================================================
      for(cardinal i = 0;i < layers;i++) {
         oldBytes[i]       = bytes[i];
         oldPackets[i]     = packets[i];
         oldPacketsLost[i] = packetsLost[i];
      }
      oldTS      = now;
      Thread::delay(1000000,true);
   }


   // ====== Clean up =======================================================
cleanUp:
   cleanUp(0);
}
